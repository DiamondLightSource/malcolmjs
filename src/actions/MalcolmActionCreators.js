/*eslint-env es6*/

/**
 * Created by twi18192 on 02/03/16.
 */

import AppDispatcher from '../dispatcher/appDispatcher.js';
import appConstants from '../constants/appConstants.js';
import MalcolmUtils from '../utils/MalcolmUtils';
import params from '../utils/queryParams';


class CMalcolmActionCreators {

constructor()
  {
    this.deviceId = params.getDeviceId();
    // TODO: update this when path changes with something like:
    //this.setQueryParams({path: ["P", "Q"]});
  }

initialiseFlowChart(requestedData)
  {

  /* Try sending an initialise flowChart start action here */

  /*
   AppDispatcher.handleAction({
   actionType: appConstants.INITIALISE_FLOWCHART_START,
   item      : 'initialise flowChart start'
   });
   */

  // NB: bind in this context, forces a pre-specified initial argument of requestedData.
  // The specified function argument becomes registered with WebSocketClient as a OnOpen callback,
  // which initiates the first Get of the top level layout information.
  MalcolmUtils.initialiseFlowChart(this.malcolmGet.bind(null, requestedData));

  //window.alert("initialisation finished?");

  //AppDispatcher.handleAction({
  //  actionType: appConstants.INITIALISE_FLOWCHART_END,
  //  item: "initialise flowChart end"
  //});

  /* Testing subscribe */
  //MalcolmUtils.initialiseFlowChart(this.malcolmSubscribe.bind(null, requestedData));

  }

/**
 * malcolmCall
 * @param blockName
 * @param method
 * @param args
 */
malcolmCall(blockName, method, args)
  {
  let requestedDataToWritePath;

  AppDispatcher.handleAction({
    actionType: appConstants.MALCOLM_CALL_PENDING,
    item      : {
      requestedDataToWrite: {
        blockName: blockName,
        method   : method,
        args     : args
      }
    }
  });

  function malcolmCallSuccess(id, responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_CALL_SUCCESS,
      item      : {
        responseMessage     : responseMessage,
        requestedDataToWrite: {
          blockName: blockName,
          method   : method,
          args     : args
        }
      }
    });
    }

  function malcolmCallFailure(responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_CALL_FAILURE,
      item      : {
        responseMessage     : responseMessage,
        requestedDataToWrite: {
          blockName: blockName,
          method   : method,
          args     : args
        }
      }
    });
    }


  requestedDataToWritePath = [actionCreators.deviceId, blockName];

  MalcolmUtils.malcolmCall(requestedDataToWritePath, method, args, malcolmCallSuccess, malcolmCallFailure);

  }


/**
 * malcolmPut()
 * @param blockName - can be empty, in which case it is defaulted to the device ID (e.g. ["P"])
 * @param endpoint - array of path ids, e.g. ["P", "layout", "value"]
 * @param value    - object to write, e.g. {mri:array[n], name:array[n], visible:array[n], x:array[n], y:array[n], typeid:array[n]}
 *
 */
malcolmPut(blockName, endpoint, value)
  {
  let requestedDataToWritePath;

  function malcolmPutSuccess(id, responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_PUT_SUCCESS,
      item      : {
        responseMessage     : responseMessage,
        requestedDataToWrite: {
          blockName: blockName,
          endpoint : endpoint,
          value    : value
        }
      }
    });
    }

  function malcolmPutFailure(responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_PUT_FAILURE,
      item      : {
        responseMessage     : responseMessage,
        requestedDataToWrite: {
          blockName: blockName,
          endpoint : endpoint,
          value    : value
        }

      }
    });
    }

  /**
   * Output needs to be of the format:
   * {"typeid":"malcolm:core/Put:1.0","id":0,"path":["P:FMC","outPwrOn","value"],"value":"On"}
   * accepted

   Memory jogger...
   check if the variable has a truthy value or not via simple test:
   if( value ) {}
   will evaluate to true if value is not: null, undefined, NaN, empty string (""), 0, false
   */
  if (blockName)
    {
    requestedDataToWritePath = [blockName];
    }
  else
    {
    requestedDataToWritePath = [actionCreators.deviceId];
    }

  MalcolmUtils.malcolmPut(requestedDataToWritePath, endpoint, value, malcolmPutSuccess, malcolmPutFailure);

  }

malcolmPost(blockName, endpoint, parameters)
  {
  let requestedDataToWritePath;

  function malcolmPostSuccess(id, responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_POST_SUCCESS,
      item      : {
        responseMessage     : responseMessage,
        requestedDataToWrite: {
          blockName : blockName,
          endpoint  : endpoint,
          parameters: parameters
        }
      }
    });
    }

  function malcolmPostFailure(responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_POST_FAILURE,
      item      : {
        responseMessage     : responseMessage,
        requestedDataToWrite: {
          blockName : blockName,
          endpoint  : endpoint,
          parameters: parameters
        }

      }
    });
    }

  /**
   * Output needs to be of the format:
   * {"typeid":"malcolm:core/Put:1.0","id":0,"path":["P:FMC","outPwrOn","value"],"value":"On"}
   */
  if (blockName)
    {
    requestedDataToWritePath = [blockName];
    }
  else
    {
    requestedDataToWritePath = [actionCreators.deviceId];
    }

  let value = "";
  MalcolmUtils.malcolmPut(requestedDataToWritePath, endpoint, value, malcolmPostSuccess, malcolmPostFailure);

  }

malcolmGet(requestedData)
  {
  function malcolmGetSuccess(id, responseMessage)
    {
    // Dispatch MALCOLM_GET_SUCCESS to all subscribers.
    AppDispatcher.handleAction(
      {
        actionType: appConstants.MALCOLM_GET_SUCCESS,
        item      : {
          responseMessage: responseMessage,
          index          : id,
          requestedData  : requestedData
        }
      });
    }

  // Dispatch MALCOLM_GET_FAILURE to all subscribers.
  function malcolmGetFailure(responseMessage)
    {
    AppDispatcher.handleAction(
      {
        actionType: appConstants.MALCOLM_GET_FAILURE,
        item      : {
          responseMessage: responseMessage,
          requestedData  : requestedData
        }
      })
    }

  // Note: This is the callback function which handles the
  // subscription return.
  let zVisibilitySubscribe = function (id, zVisibility)
    {
    //console.log('MalcolmActionCreators: zVisibility...')
    //console.log(zVisibility);
    // zVisibility represents the top level object for the device
    // which includes layout and visibility details:
    // zVisibility.layout.value.visible[]
    // IJG 31/10/16

    /* This is to return Z:VISIBILITY to the GUI, but I still
     need to subscribe to all the blocks, so a for loop is
     part of the callback too
     */

    // //malcolmGetSuccess(id, zVisibility);
    //if (zVisibility.hasOwnProperty('layout'))
    //  {
    //  blockCollection.createBlockItemsFromSchema(zVisibility);
    //  }


    // Subscribe to the visibility list.
    //actionCreators.malcolmSubscribe(actionCreators.deviceId,["layout","value","visible"]);

    // todo: Consider the following subscriptions...
    //actionCreators.malcolmSubscribe(actionCreators.deviceId,["layout","value","x"]);
    //actionCreators.malcolmSubscribe(actionCreators.deviceId,["layout","value","y"]);

    //==========================================================================================
    // TODO:
    // The old Protocol_1 attributes pattern do not fit with Protocol_2.
    // Instead we need to interrogate each block in the upper list (from Get),
    // by Get or Subscribe to each block, then assemble the attributes into a local store.
    // It begs the question as to why there is an attributeStore instead of each block item holding
    // its own attributes?
    // Ian Gillingham - October 2016.
    //==========================================================================================


    };

  /**
   * testMalcolmGetSuccess: Callback to top level layout response
   *
   * @param id
   * @param responseMessage
   */
  let testMalcolmGetSuccess = function (id, responseMessage)
    {
    //console.log(`MalcolmActionCreators: testMalcolmGetSuccess: responseMessage...  topLevelGetId = ${actionCreators.topLevelGetId}`)
    //console.log(responseMessage);

    /* Fetch Z:VISIBILITY, and then subscribe to all visible
     attributes in Z:VISIBILITY
     */

    // Instruct MalcolmUtils to get the visibility information for all blocks on the remote instrument.
    // Was: MalcolmUtils.malcolmGet('Z:VISIBILITY', zVisibilitySubscribe, malcolmGetFailure);
    MalcolmUtils.malcolmGet([actionCreators.deviceId], zVisibilitySubscribe, malcolmGetFailure);

    };


  //console.log('MalcolmActionCreators: malcolmGet: Requested data...');
  //console.log(requestedData);
  //console.log('MalcolmActionCreators: malcolmGet: deviceId...');
  //console.log(actionCreators.deviceId);

  if (requestedData === actionCreators.deviceId)
    {

    /*
     ===========================================================================
     This is the initial malcolmGet call that returns the top level device block (once called Z);
     It is then passed a callback function that loops through all the contained
     attributes.blocks.value and fetches each block in that list
     ===========================================================================
     */

    actionCreators.topLevelGetId = MalcolmUtils.malcolmGet(requestedData, testMalcolmGetSuccess, malcolmGetFailure);
    //console.log(`MalcolmActionCreators: malcolmGet(): Issue Request for data:  topLevelGetId = ${actionCreators.topLevelGetId}`);
    }
  else
    {
    actionCreators.topLevelGetId = MalcolmUtils.malcolmGet(requestedData, malcolmGetSuccess, malcolmGetFailure);
    //console.log(`MalcolmActionCreators: malcolmGet() [requestedData !== actionCreators.deviceId]: Issue Request for data:  topLevelGetId = ${actionCreators.topLevelGetId}`);
    }

  }

/**
 * @name malcolmSubscribe
 * @param {string} blockName - MRI name of block (e.g. "P:OUTENC3")
 * @param {[string]} attribute - array of attribute tree to subscribe to
 * @param {string} subtype - actionType response - defaults to MALCOLM_SUBSCRIBE_SUCCESS.
 */
malcolmSubscribe(blockName, attribute, subtype = appConstants.MALCOLM_SUBSCRIBE_SUCCESS)
  {
  /**
   * alt_response wil be defaulted to MALCOLM_SUBSCRIBE_SUCCESS if the subtype argument is not supplied.
   * It provides closure within the success and failure callback functions and defines what Flux message (actionType)
   * will be dispatched.
   * IJG 7 March 2017
   *
   * @type {string}
   */
  let alt_response = subtype;

  function malcolmSubscribeSuccess(id, responseMessage)
    {
    //console.log(`MalcolmActionCreators.malcolmSubscribeSuccess(): id = ${id}`);
    // TODO: Lookup which return ID to adopt for the promise.
    AppDispatcher.handleAction({
      actionType: alt_response,
      item      : {
        responseMessage: responseMessage,
        index          : id,
        requestedData  : {
          blockName: blockName,
          attribute: attribute
        }
      }
    });
    }

  function malcolmSubscribeFailure(responseMessage)
    {
    AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_SUBSCRIBE_FAILURE,
      item      : {
        responseMessage: responseMessage,
        requestedData  : {
          blockName: blockName,
          attribute: attribute
        }
      }
    });
    }

  //let requestedAttributeDataPath = [actionCreators.deviceId, blockName, "attributes", attribute];
  let requestedAttributeDataPath = null;
  if (attribute.length > 0)
    {
    requestedAttributeDataPath = [blockName].concat(attribute);
    }
  else
    {
    requestedAttributeDataPath = [blockName];
    }


  //console.log(`MalcolmActionCreators.malcolmSubscribe(): ${requestedAttributeDataPath}`);
  let id = MalcolmUtils.malcolmSubscribe(requestedAttributeDataPath, malcolmSubscribeSuccess, malcolmSubscribeFailure);
  return (id);
  }

/**
 * @name  malcolmAttributeValueEdited
 * @param blockName
 * @param attributePath
 * @param newValue
 * @description  Called by a UI component when a block attribute value is changed by the user.
 *               Dispatches a message to all stores, so that one of the can take the appropriate action,
 *               such as Pushing the new value to the device.
 */
malcolmAttributeValueEdited(blockName, attributePath, newValue)
  {
  AppDispatcher.handleAction({
      actionType: appConstants.MALCOLM_BLOCK_ATTRIBUTE_EDITED,
      item      : {
        blockName     : blockName,
        attribute_path: attributePath,
        newValue      : newValue
      }
    }
  );
  }

getdeviceId()
  {
  return ( actionCreators.deviceId );
  }
}

const MalcolmActionCreators = new CMalcolmActionCreators();
const actionCreators        = MalcolmActionCreators;


export default MalcolmActionCreators;
