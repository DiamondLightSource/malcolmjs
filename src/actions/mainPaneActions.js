/**
 * Created by twi18192 on 25/08/15.
 */

import AppDispatcher from '../dispatcher/appDispatcher.js';
import appConstants  from '../constants/appConstants.js';

class MainPaneActions {
  constructor()
    {

    }

  toggleFooter1(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.FOOTER_TOGGLE,
      item: item
    })
  }

  toggleConfigPanel(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.CONFIG_TOGGLE,
      item: item
    })
  }
  toggleFavPanel(item){
    AppDispatcher.handleViewAction({
      actionType: appConstants.FAV_TOGGLE,
      item: item
    })
  }
}

const mainPaneActions = new MainPaneActions();

export default mainPaneActions;
