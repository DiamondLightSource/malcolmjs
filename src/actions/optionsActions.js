/**
 * Created by Ian Gillingham 29 August 2017.
 */

import AppDispatcher from '../dispatcher/appDispatcher';
import appConstants from '../constants/appConstants';

let optionsActions = {

  GraphicsStyleSelect: function(gstyle){
    AppDispatcher.handleViewAction({
      actionType: appConstants.OPTIONS_ACTION_GSTYLE,
      item:gstyle
    })
  },
  RightPanePinnedSelect: function(rightPinned){
    AppDispatcher.handleViewAction({
      actionType: appConstants.OPTIONS_ACTION_RIGHT_PINNED,
      item: rightPinned
    })
  }
};

export default optionsActions;
