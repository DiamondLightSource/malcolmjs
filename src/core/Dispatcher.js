/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
