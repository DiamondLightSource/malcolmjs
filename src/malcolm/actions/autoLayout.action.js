import ELK from 'elkjs/lib/elk.bundled';
import { malcolmPutAction, malcolmSetFlag } from '../malcolmActionCreators';

const calculateHeight = (layoutEngine, mri) => {
  const zoomFactor = layoutEngine.diagramModel.zoom / 100;
  return (
    layoutEngine.getNodeDimensions(layoutEngine.diagramModel.nodes[mri])
      .height / zoomFactor
  );
};

const runAutoLayout = () => (dispatch, getState) => {
  const state = getState();
  const { blocks } = state.malcolm.layout;
  const { parentBlock, mainAttribute } = state.malcolm;
  const { links } = state.malcolm.layoutEngine.diagramModel;

  const graph = {
    id: 'root',
    layoutOptions: {
      'elk.algorithm': 'layered',
      separateConnectedComponents: true,
      'spacing.componentComponent': 70,
      spacing: 100,
      'spacing.nodeNodeBetweenLayers': 70,
      'elk.edgeRouting': 'SPLINES',
    },
    children: blocks.map(b => ({
      id: b.mri,
      width: 120,
      height: calculateHeight(state.malcolm.layoutEngine, b.mri),
      x: b.position.x,
      y: b.position.y,
      properties: {
        portConstraints: 'FIXED_ORDER',
      },
      ports: b.ports.map((p, i) => ({
        id: `${b.mri}-${p.label}`,
        properties: {
          side: p.input ? 'WEST' : 'EAST',
          index: p.input ? b.ports.length - i : i,
        },
        width: 12,
        height: 12,
      })),
    })),
    edges: Object.values(links).map((link, i) => ({
      id: `e${i}`,
      sources: [link.sourcePort.id],
      targets: [link.targetPort.id],
    })),
  };

  // The JSON graph can be visualised here https://rtsys.informatik.uni-kiel.de/elklive/json.html
  console.log(JSON.stringify(graph));
  console.log(
    `Graph can be visualised at https://rtsys.informatik.uni-kiel.de/elklive/json.html`
  );

  const elk = new ELK();
  return elk
    .layout(graph)
    .then(graphLayout => {
      console.log('completed auto layout');
      const centerX =
        graphLayout.children.reduce((prev, next) => prev + next.x, 0) /
        graphLayout.children.length;
      const centerY =
        graphLayout.children.reduce((prev, next) => prev + next.y, 0) /
        graphLayout.children.length;

      // we only need to send an update for the visible blocks
      const updatedLayout = {
        name: blocks.map(b => b.name),
        visible: blocks.map(b => b.visible),
        mri: blocks.map(b => b.mri),
        x: blocks
          .map(b => b.mri)
          .map(mri => graphLayout.children.find(c => c.id === mri).x - centerX),
        y: blocks
          .map(b => b.mri)
          .map(mri => graphLayout.children.find(c => c.id === mri).y - centerY),
      };

      dispatch(malcolmSetFlag([parentBlock, mainAttribute], 'pending', true));
      dispatch(
        malcolmPutAction([parentBlock, mainAttribute, 'value'], updatedLayout)
      );
    })
    .catch(err => {
      console.log('error creating auto layout');
      console.log(err);
    });
};

export default {
  runAutoLayout,
};
