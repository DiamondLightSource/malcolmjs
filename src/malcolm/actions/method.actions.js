import { MalcolmAttributeData } from '../malcolm.types';

export const buildMethodUpdate = (id, data) => ({
  type: MalcolmAttributeData,
  payload: {
    id,
    ...data,
    isMethod: true,
    delta: true,
    pending: false,
  },
});

export default {
  buildMethodUpdate,
};
