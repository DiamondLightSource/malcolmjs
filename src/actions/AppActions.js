/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Dispatcher from '../core/Dispatcher';
import ActionTypes from '../constants/ActionTypes';
import ExecutionEnvironment from 'react/lib/ExecutionEnvironment';
import http from 'superagent';

export default {

  navigateTo(path, options) {
    if (ExecutionEnvironment.canUseDOM) {
      if (options && options.replace) {
        window.history.replaceState({}, document.title, path);
      } else {
        window.history.pushState({}, document.title, path);
      }
    }

    Dispatcher.handleViewAction({
      actionType: ActionTypes.CHANGE_LOCATION,
      path
    });
  },

  loadPage(path, cb) {
    Dispatcher.handleViewAction({
      actionType: ActionTypes.LOAD_PAGE,
      path
    });

    http.get('/api/page' + path)
      .accept('application/json')
      .end((err, res) => {
        Dispatcher.handleServerAction({
          actionType: ActionTypes.LOAD_PAGE,
          path,
          err,
          page: res.body
        });
        if (cb) {
          cb();
        }
      });
  }

};
