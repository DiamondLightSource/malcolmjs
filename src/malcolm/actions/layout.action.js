import {
  MalcolmSelectPortType,
  MalcolmMakeBlockVisibleType,
  MalcolmShowBinType,
  MalcolmInLayoutDeleteZoneType,
  MalcolmResetPortsType,
} from '../malcolm.types';
import {
  malcolmPutAction,
  malcolmSetFlag,
  malcolmSelectBlock,
} from '../malcolmActionCreators';
import blockUtils from '../blockUtils';
import { snackbarState } from '../../viewState/viewState.actions';
import { idSeparator } from '../../layout/layout.component';
import { sinkPort } from '../malcolmConstants';

const findPort = (blocks, id) => {
  const path = id.split(idSeparator);
  const block = blocks.find(b => b.mri === path[0]);
  const port = block.ports.find(p => p.label === path[1]);

  return port;
};

const selectPortInClient = (portId, start) => ({
  type: MalcolmSelectPortType,
  payload: {
    portId,
    start,
  },
});

export const selectPort = (portId, start) => (dispatch, getState) => {
  if (!portId) {
    if (getState().malcolm.layoutState.startPortForLink) {
      dispatch({
        type: MalcolmResetPortsType,
        payload: {},
      });
    }
    return;
  }

  const port = findPort(getState().malcolm.layout.blocks, portId);

  if (port && port.input && port.value !== port.tag) {
    // Reset the ports
    dispatch({
      type: MalcolmResetPortsType,
      payload: {},
    });

    // Display an error message
    dispatch(
      snackbarState(
        true,
        `Error: Existing link to input port ${
          port.label
        }, delete existing link first.`
      )
    );
  } else {
    dispatch(selectPortInClient(portId, start));

    const { layoutState, layout } = getState().malcolm;
    const { startPortForLink, endPortForLink } = layoutState;

    if (startPortForLink && endPortForLink) {
      const startPort = findPort(layout.blocks, startPortForLink);
      const endPort = findPort(layout.blocks, endPortForLink);

      const outputPort = startPort.input ? endPort : startPort;

      // 2. update on the server
      const path = [
        ...(startPort.input ? startPortForLink : endPortForLink).split(
          idSeparator
        ),
        'value',
      ];

      dispatch(malcolmSetFlag(path, 'pending', true));
      dispatch(malcolmPutAction(path, outputPort.tag));
    }
  }
};

const makeBlockVisible = (mriList, positionList, visibleList) => (
  dispatch,
  getState
) => {
  let mri;
  let position;
  let visible;
  if (!(mriList instanceof Array)) {
    mri = [mriList];
    position = [positionList];
    visible = [visibleList === undefined ? true : visibleList];
  } else {
    mri = [...mriList];
    position = [...positionList];
    visible = [...visibleList];
  }

  const state = getState().malcolm;
  const positionRelativeToCenter = position.map(pos => ({
    x: pos.x - state.layoutState.layoutCenter.x,
    y: pos.y - state.layoutState.layoutCenter.y,
  }));

  mri.forEach((blockMri, index) => {
    dispatch({
      type: MalcolmMakeBlockVisibleType,
      payload: {
        mri: blockMri,
        position: positionRelativeToCenter[index],
        visible: visible[index],
      },
    });
  });

  const blockName = state.parentBlock;
  const attributeName = state.mainAttribute;
  const layoutAttribute = blockUtils.findAttribute(
    state.blocks,
    blockName,
    attributeName
  );

  if (layoutAttribute && mri.length > 0) {
    const visibleBlockIndices = mri.map(blockMri =>
      layoutAttribute.raw.value.mri.findIndex(val => val === blockMri)
    );
    const updateLayout = {
      name: visibleBlockIndices.map(
        index => layoutAttribute.raw.value.name[index]
      ),
      visible,
      mri,
      x: positionRelativeToCenter.map(pos => pos.x),
      y: positionRelativeToCenter.map(pos => pos.y),
    };

    dispatch(malcolmSetFlag([blockName, attributeName], 'pending', true));
    dispatch(
      malcolmPutAction([blockName, attributeName, 'value'], updateLayout)
    );
  }
};

const deleteBlocks = () => (dispatch, getState) => {
  const state = getState().malcolm;
  const { selectedBlocks } = state.layoutState;

  makeBlockVisible(
    selectedBlocks,
    selectedBlocks.map(() => state.layoutState.layoutCenter),
    selectedBlocks.map(() => false)
  )(dispatch, getState);
  selectedBlocks.forEach(b => {
    dispatch(malcolmSelectBlock(b, false));
  });
};

const deleteLinks = () => (dispatch, getState) => {
  const state = getState().malcolm;

  state.layoutState.selectedLinks.forEach(linkId => {
    const [, , blockMri, linkAttr] = linkId.split(idSeparator);
    const portAttribute = blockUtils.findAttribute(
      state.blocks,
      blockMri,
      linkAttr
    );
    if (portAttribute) {
      const portNullValue = portAttribute.raw.meta.tags
        .find(t => t.indexOf(sinkPort) > -1)
        .split(':')
        .slice(-1)[0];
      dispatch(malcolmSetFlag([blockMri, linkAttr], 'pending', true));
      dispatch(malcolmPutAction([blockMri, linkAttr], portNullValue));
    }
  });
};

const showLayoutBin = show => ({
  type: MalcolmShowBinType,
  payload: {
    visible: show,
  },
});

const mouseInsideDeleteZone = insideZone => ({
  type: MalcolmInLayoutDeleteZoneType,
  payload: {
    insideZone,
  },
});

export default {
  selectPort,
  makeBlockVisible,
  showLayoutBin,
  mouseInsideDeleteZone,
  deleteBlocks,
  deleteLinks,
};
