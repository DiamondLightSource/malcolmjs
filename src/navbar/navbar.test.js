import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import configureStore from 'redux-mock-store';
import NavBar from './navbar.component';
import { openParentPanelType } from '../viewState/viewState.actions';

const mockStore = configureStore();
const dispatch = jest.fn();

describe('NavBar', () => {
  let shallow;
  let mount;
  let state;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    mount = createMount();

    state = {
      viewState: {
        openParentPanel: true,
        openChildPanel: true,
      },
      malcolm: {
        blocks: {},
        navigation: {
          navigationLists: [
            { path: 'PANDA', children: [] },
            { path: 'layout', children: [] },
            { path: 'PANDA:SEQ1', children: [] },
          ],
          rootNav: {
            path: '',
            children: ['PANDA', 'PANDA:SEQ1'],
          },
        },
      },
    };
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <NavBar store={mockStore(state)} dispatch={dispatch} />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('openParent calls the open method', () => {
    state.viewState.openParentPanel = false;
    const store = mockStore(state);

    const wrapper = mount(<NavBar store={store} />);

    wrapper
      .find('button')
      .at(0)
      .simulate('click');

    const actions = store.getActions();
    expect(actions.length).toEqual(1);
    expect(actions[0].type).toBe(openParentPanelType);
    expect(actions[0].openParentPanel).toBeTruthy();
  });
});
