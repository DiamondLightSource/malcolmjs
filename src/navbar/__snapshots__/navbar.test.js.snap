// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders correctly 1`] = `
<WithStyles(AppBar)
  className="Connect-NavBar--appBar-1 Connect-NavBar--appBarShift-2 Connect-NavBar--appBarShift-left-3"
>
  <WithStyles(Toolbar)
    disableGutters={false}
  >
    <WithStyles(IconButton)
      className="Connect-NavBar--menuButton-5 Connect-NavBar--hide-6"
      onClick={[Function]}
    >
      <pure(Menu) />
    </WithStyles(IconButton)>
    <div
      className="Connect-NavBar--title-7 Connect-NavBar--titleShift-8"
    >
      <WithStyles(NavControl)
        key=".blocks"
        nav={
          Object {
            "children": Array [
              "PANDA",
              "PANDA:SEQ1",
            ],
            "path": "",
          }
        }
        navigateToChild={[Function]}
      />
      <WithStyles(NavControl)
        key="PANDA"
        nav={
          Object {
            "children": Array [],
            "path": "PANDA",
          }
        }
        navigateToChild={[Function]}
      />
      <WithStyles(NavControl)
        key="layout"
        nav={
          Object {
            "children": Array [],
            "path": "layout",
          }
        }
        navigateToChild={[Function]}
      />
      <WithStyles(NavControl)
        key="PANDA:SEQ1"
        nav={
          Object {
            "children": Array [],
            "path": "PANDA:SEQ1",
          }
        }
        navigateToChild={[Function]}
      />
    </div>
  </WithStyles(Toolbar)>
</WithStyles(AppBar)>
`;
